ballista/core/src/physical_optimizer/task_group.rs

```diff
@@ -110,42 +100,22 @@ impl OptimizeTaskGroup {
                 Ok(Transformed::No(node))
             }
         } else {
-            let is_union = node.as_any().is::<UnionExec>();
-            let all_children_are_coalesce = node
-                .children()
-                .iter()
-                .all(|child| child.as_any().is::<CoalesceTasksExec>());
-
-            if (is_union || is_hash_join_no_partitioning(node.as_ref()))
-                && all_children_are_coalesce
-            {
-                let new_children =
-                    children.iter().flat_map(|child| child.children()).collect();
-
-                let new_plan: Arc<dyn ExecutionPlan> = Arc::new(CoalesceTasksExec::new(
-                    node.clone().with_new_children(new_children)?,
-                    self.partitions.clone(),
-                ));
-
-                return Ok(Transformed::Yes(new_plan));
-            }
-
             Ok(Transformed::No(node))
         }
     }
 }
 
-fn is_hash_join_no_partitioning(node: &dyn ExecutionPlan) -> bool {
-    // only push down hash join if the input is unpartitioned
-    // (so parent nodes won't rely on the output partitioning)
-    if let Some(hash_join) = node.as_any().downcast_ref::<HashJoinExec>() {
-        return matches!(
-            hash_join.output_partitioning(),
-            datafusion::physical_plan::Partitioning::UnknownPartitioning(_)
-        );
-    }
-    false
-}
+// fn is_hash_join_no_partitioning(node: &dyn ExecutionPlan) -> bool {
```

Comment:
@Dandandan Slightly hazy on this but pretty sure we always we want to coalesce after `HashJoinExec` even if it's partitioned.

Reply:
I think you're right.
The idea was I think we can't pushdown `HashJoinExec` for non-partitioned data, but it should already run in another stage at this ppoint if the child requires a hash-partitioning.